File: Typescript/easy/Exercism/word-count.ts
// # Word Count

// Given a phrase, count the occurrences of each word in that phrase.

// For example for the input `"olly olly in come free"`

// ```text
// olly: 2
// in: 1
// come: 1
// free: 1
// ```

export default class Words {
    public count(phrase: string): Map<string, number> {
        phrase = phrase.trim().replace(/\n|\t+|  +/g, ' ');
        const words = phrase.split(' ');
        const wordFrequency = new Map<string, number>();
        
        for(let word of words){
            word = word.toLocaleLowerCase();
            if(wordFrequency.has(word)){
                wordFrequency.set(word, wordFrequency.get(word)! + 1);
            } else {
                wordFrequency.set(word, 1);
            }
        }

        return wordFrequency;
    }
}

File: Java/medium/Exercism/word-count/src/main/java/WordCount.java
import java.util.HashMap;
import java.util.Map;

class WordCount {
    public Map<String, Integer> phrase(String input) {
        Map<String, Integer> wordCount = new HashMap<>();

        // Normalize input and split into words
        String[] words = input.toLowerCase().trim().split("[^a-z0-9']+");

        // Count occurrences of each word
        for (String word : words) {
            // remove leading and trailing single quotes if they exist
            if (word.startsWith("'")) {
                word = word.substring(1);
            }
            if (word.endsWith("'")) {
                word = word.substring(0, word.length() - 1);
            }

            // Skip empty words
            if (word.isEmpty()) {
                continue;
            }

            if (wordCount.containsKey(word)) {
                wordCount.put(word, wordCount.get(word) + 1);
            } else {
                wordCount.put(word, 1);
            }
        }

        return wordCount;
    }
}
